python_modules/dagster/dagster/_config/structured_config.py

```diff
@@ -132,7 +156,55 @@ def create_object_to_pass_to_user_code(self, context) -> Any:  # pylint: disable
         Default behavior for new class-based resources is to return itself, passing
         the actual resource object to user code.
         """
-        return self
+        return cast(T, self)
+
+    @classmethod
+    def partial(cls, **kwargs) -> "PartialResource[Self]":
+        """
+        Returns a partially initialized copy of the resource, with remaining config fields
+        set at runtime.
+        """
+        return PartialResource(cls, data=kwargs)
+
+    # Python descriptor
+    # https://docs.python.org/3/howto/descriptor.html
+    # Used to adjust the types of resource inputs and outputs, e.g. resource dependencies can be passed in
+    # as PartialResources or Resources, but will always be returned as Resources
+    def __get__(self, obj: "Resource", __owner: Any) -> Self:
+        ...
+
+    def __set__(self, obj: Optional[object], value: Union[T, "PartialResource[T]"]) -> None:
+        ...
+
+
+class PartialResource(Generic[T], ResourceDefinition, MakeConfigCacheable):
+    data: Dict[str, Any]
+    resource_cls: Type[Resource[T]]
+
+    def __init__(self, resource_cls: Type[Resource[T]], data: Dict[str, Any]):
+        check.invariant(data == {}, "PartialResource currently does not support config fields")
+
+        Config.__init__(self, data=data, resource_cls=resource_cls)
```

Comment:
It seems very strange to directly call the `__init__` for a class two notches up in the inheritance chain